### DISCUSSION TITLE
Reapply rules

### DISCUSSION THREAD

**Discussion by exic** [asmvik/yabai#2119] [Q&A] [answered]
Sometimes when attaching/removing external displays, some of my apps end up on different spaces than I want them to. Is there an easy way to reapply all of my rules that put them to my labelled spaces as before, or do I need to label/extract them to write something that removes/readds them?

If I just restart yabai it does what I want for those "rule-managed" windows but then other windows might get moved around unexpectedly..

**Comment by asmvik** [comment]
Currently you'd need to label rules and factor them into a separate script that manages lifetime such as removing and adding them.
I would be open for adding a  command like.. yabai -m rule --reevaluate that could take an index/label/or nothing (meaning all rules) will run.

**Comment by exic** [comment]
I have now labelled my placement rules and moved them to an external script, will see if that works for me.

**Comment by asmvik** [comment, marked as answer]
Implemented on master.

https://github.com/koekeishiya/yabai/issues/2121

**Reply by exic** [reply]
That was fast, thanks a lot!

**Reply by asmvik** [reply]
This is going to change a bit before it goes live in a release, please track #2123. 
In short, to re-apply an existing rule to currently known windows in the future:

`yabai -m rule --apply [<RULE_SEL> | <ARGUMENTS>]`

if no argument is given, it will apply all registered rules.
If an index or label is given, that particular rule will apply.

You can also decide to provide the arguments directly, e.g:

`yabai -m rule --apply app="^Safari$" manage=off grid=1:1:0:0:1:1 opacity=1.0`

This is implemented on master.
